src/libraries/System.Net.Security/src/System/Net/Security/Pal.OSX/SafeDeleteSslContext.cs

```diff
@@ -143,6 +144,12 @@ protected override void Dispose(bool disposing)
             {
                 if (null != _sslContext)
                 {
+                    lock (_sslContext)
+                    {
+                        _inputBuffer.Dispose();
+                        _outputBuffer.Dispose();
+                    }
+
                     _sslContext.Dispose();
                     _sslContext = null;
```

Comment:
This code smells like a race condition to me.

If a preemption occurs right after `_sslContext` is unlocked and before `_sslContext` is disposed, and another thread disposes of `_sslContext` and sets it to `null`, whenever the thread that disposed of the buffers comes along again, it'll call `Dispose()` on a null reference.

Another thing that I'm worried here is that if one thread locks `_sslContext`, disposes the buffers, unlocks `_sslContext`, disposes of `_sslContext`, while another thread was waiting for the `_sslContext` lock to be available again, this might cause issues.  (I'm not intimately familiar with how lifetimes work in .NET.)

Reply:
> This code smells like a race condition to me.

Agreed; it seems like you might end up getting `lock(null)` as well.

> Another thing that I'm worried here is that if one thread locks `_sslContext`, disposes the buffers, unlocks `_sslContext`, disposes of `_sslContext`, while another thread was waiting for the `_sslContext` lock to be available again, this might cause issues. (I'm not intimately familiar with how lifetimes work in .NET.)

This should work fine; disposal is separate from an object's ability to be used as a lock.

Reply:
updated. Since other code also uses _sslContext, I removed the null assignment as it does not seems necessary. (We also use IsInvalid in some places)